flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,557 @@
+        assertThat(informationProvider.getJarFile().isPresent(), is(false));
+    }
+
+    @Test
+    public void testCreateWithUserLibDir() throws FlinkException {
+        final PackagedProgramRetriever retriever =
+                PackagedProgramRetrieverImpl.create(
+                        classpathProvider.getUserDirectoryWithOneEntryClass(),
+                        null,
+                        classpathProvider.getJobClassName(),
+                        new String[0],
+                        new Configuration());
+
+        // the right information is picked up without any error
+        assertThat(
+                retriever.getPackagedProgram().getMainClassName(),
+                is(classpathProvider.getJobClassName()));
+    }
+
+    @Test
+    public void testJobGraphRetrieval()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final int parallelism = 42;
+        final JobID jobId = new JobID();
+
+        final Configuration configuration = new Configuration();
+        configuration.setInteger(CoreOptions.DEFAULT_PARALLELISM, parallelism);
+        configuration.set(PipelineOptionsInternal.PIPELINE_FIXED_JOB_ID, jobId.toHexString());
+
+        final String expectedSuffix = "suffix";
+        final PackagedProgramRetriever retriever =
+                PackagedProgramRetrieverImpl.create(
+                        null,
+                        classpathProvider.getTestJobClassName(),
+                        classpathProvider.getTestJobArgs(expectedSuffix),
+                        new Configuration());
+
+        final JobGraph jobGraph = retrieveJobGraph(retriever, configuration);
+
+        assertThat(
+                jobGraph.getName(),
+                is(classpathProvider.getTestJobClassName() + "-" + expectedSuffix));
+        assertThat(jobGraph.getSavepointRestoreSettings(), is(SavepointRestoreSettings.none()));
+        assertThat(jobGraph.getMaximumParallelism(), is(parallelism));
+        assertThat(jobGraph.getJobID(), is(jobId));
+    }
+
+    @Test
+    public void testJobGraphRetrievalFromJar()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final String expectedSuffix = "suffix";
+        final PackagedProgramRetriever retrieverUnderTest =
+                PackagedProgramRetrieverImpl.create(
+                        classpathProvider.getUserDirectoryWithTestJob(),
+                        null,
+                        null,
+                        classpathProvider.getTestJobArgs(expectedSuffix),
+                        new Configuration());
+
+        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, new Configuration());
+
+        assertThat(
+                jobGraph.getName(),
+                is(classpathProvider.getTestJobClassName() + "-" + expectedSuffix));
+    }
+
+    @Test
+    public void testJobGraphRetrievalJobClassNameHasPrecedenceOverClasspath()
```

Comment:
I left the test here for now for transparency reasons. I think we should remove it but make the job fail in case the job class is not even provided on the system class path.

Reply:
Yes. The legacy `ClassPathPackagedProgramRetriever.Builder#setJarsOnClassPath` is never used in production code. And it seems not have some benefits to specify the jar name on classpath instead of entry class.

+1 to remove this test.

Reply:
It doesn't matter where the setter is used in production or not, the purpose of the test is to ensure behavior despite _changes_ in production code (i.e., if the setter were used).

Where else do you test that, if 2 entrypoints are on the CP and the user specifies the entrypoint class, that the user argument is honored?

Reply:
This test only succeeds in the old implementation because we're not checking whether the given job class name is on the system classpath. But @zentol is right with checking that the parameters are acknowledged for this specific case. Although, it should be moved to `PackagedProgramTest`, I guess. But I leave it here (and just rename it) because I feel that that refactoring the testcases in this matter is out of scope of this PR.